go/libraries/doltcore/rowconv/row_converter.go

```diff
@@ -68,9 +77,34 @@ func NewRowConverter(ctx context.Context, vrw types.ValueReadWriter, mapping *Fi
 	return &RowConverter{mapping, false, convFuncs}, nil
 }
 
-// Convert takes a row maps its columns to their destination columns, and performs any type conversion needed to create
-// a row of the expected destination schema.
+// panicOnDuplicateMappings checks if more than one input field is mapped to the same output field.
+// Multiple input fields mapped to the same output field results in a race condition.
+func panicOnDuplicateMappings(mapping *FieldMapping) {
+	destToSrcMapping := make(map[uint64]uint64, len(mapping.SrcToDest))
+	for srcTag, destTag := range mapping.SrcToDest {
+		if _, found := destToSrcMapping[destTag]; found {
+			panic("multiple columns mapped to the same destination tag '" + types.Uint(destTag).HumanReadableString() + "'")
+		}
+		destToSrcMapping[destTag] = srcTag
+	}
+}
+
+// ConvertWithWarnings takes a row and maps its columns to their destination columns, performing any type conversions
+// needed to create a row of the expected destination schema, and logging SQL warnings for any fields that cannot
+// be cleanly converted.
+func (rc *RowConverter) ConvertWithWarnings(ctx *sql.Context, inRow row.Row) (row.Row, error) {
+	return rc.convert(ctx, inRow, true)
+}
+
+// Convert takes a row and maps its columns to their destination columns, and performs any type conversion needed to
+// create a row of the expected destination schema.
 func (rc *RowConverter) Convert(inRow row.Row) (row.Row, error) {
+	return rc.convert(nil, inRow, false)
+}
+
+// convert takes a row and maps its columns to their destination columns, automatically performing any type conversion
+// needed, and optionally logging SQL warnings on type conversion errors if the warnOnCoercionError flag is set to true
+func (rc *RowConverter) convert(ctx *sql.Context, inRow row.Row, warnOnCoercionError bool) (row.Row, error) {
```

Comment:
Little weird that ctx can be nil, but only if warn is false.

Another way to structure this would be for it to accept a `warnFn` which is optionally nil.

Reply:
oooh... great idea about `warnFn`. That made the code much cleaner! Thanks!